src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,142 @@
+    }
+
+    private int hash(K key) {
+        return key == null ? 0 : key.hashCode();
+    }
+
+    private Node<K, V>[] resize() {
+        final Node<K, V>[] oldTab = table;
+        int oldCap = ((table == null) ? 0 : table.length);
+        int oldThr = threshold;
+        int newThr;
+        int newCap;
+        if (oldCap > 0) {
+            newCap = oldCap << 1;
+            newThr = oldThr << 1;
+        } else {
+            newThr = (int) (defaultInitialCapacity * defaultLoadFactor);
+            newCap = defaultInitialCapacity;
+        }
+        threshold = newThr;
+        Node<K,V>[] newTab = (Node<K,V>[])new Node[newCap];
+        table = newTab;
+        if (oldTab != null) {
+            for (int i = 0; i < oldTab.length; i++) {
+                Node<K, V> node;
+                if ((node = oldTab[i]) != null) {
+                    oldTab[i] = null;
+                    if (node.next == null) {
+                        newTab[node.hash & (newCap - 1)] = node;
+                    } else {
+                        while (node != null) {
+                            final Node<K, V> current = node;
+                            node = node.next;
+                            putNode(current, newTab);
+                        }
+                    }
+                }
+            }
+        }
+        return newTab;
+    }
+
+    private void putNode(Node<K, V> current, Node<K, V>[] newTab) {
+        current.next = null;
+        int index;
+        if ((newTab[index = (current.hash & (newTab.length - 1))]) == null) {
+            newTab[index] = current;
+        } else {
+            Node<K, V> tn = newTab[index];
+            K key = current.key;
+            if (tn.hash == current.hash && (key == tn.key || (key != null && key.equals(tn.key)))) {
+                newTab[index].value = current.value;
+            } else {
+                while (tn.next != null) {
+                    tn = tn.next;
+                }
+                tn.next = current;
+            }
+        }
+    }
+
+    public void putVal(int hash, K key, V value) {
+        int n;
+        int index;
+        if (table == null || (n = table.length) == 0) {
+            n = (table = resize()).length;
+        }
+        if ((table[index = hash & (n - 1)]) == null) {
+            table[index] = new Node<>(hash, key, value, null);
+        } else {
+            for (Node<K, V> tn = table[index]; ; tn = tn.next) {
+                if (tn.hash == hash && (key == tn.key || (key != null && key.equals(tn.key)))) {
+                    tn.value = value;
+                    return;
+                }
+                if (tn.next == null) {
+                    tn.next = new Node<>(hash, key, value, null);
+                    break;
+                }
+            }
+        }
+        if (++size > threshold) {
+            table = resize();
+        }
+    }
 
     @Override
     public void put(K key, V value) {
+        putVal(hash(key), key, value);
```

Comment:
this method does nothing, just redirects a call to putVal. That means we can just move everything we have in putValu inside of put() and have 1 method instead of 2